File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio7.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class Ejercicio7 {

    private static final String ARCHIVO_CUENTAS = "cuentas.txt";

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Map<String, CuentaUsuario> cuentas = cargarCuentas();

        System.out.println("=== Cajero Automático Básico ===");
        String usuario = leerTextoNoVacio(scanner, "Ingrese su nombre de usuario: ");
        CuentaUsuario cuenta = cuentas.getOrDefault(usuario, new CuentaUsuario(usuario, 0.0));
        cuentas.put(usuario, cuenta);

        int opcion;
        do {
            mostrarMenu();
            opcion = leerEntero(scanner, "Seleccione una opción: ");
            switch (opcion) {
                case 1:
                    System.out.printf("Saldo actual de %s: $%.2f%n", cuenta.getUsuario(), cuenta.getSaldo());
                    break;
                case 2:
                    realizarDeposito(scanner, cuenta);
                    guardarCuentas(cuentas);
                    break;
                case 3:
                    realizarRetiro(scanner, cuenta);
                    guardarCuentas(cuentas);
                    break;
                case 4:
                    System.out.println("Gracias por usar el cajero. Hasta luego.");
                    break;
                default:
                    System.out.println("Opción inválida. Intente de nuevo.");
            }
            System.out.println();
        } while (opcion != 4);

        scanner.close();
    }

    private static void mostrarMenu() {
        System.out.println("1. Consultar saldo");
        System.out.println("2. Depositar dinero");
        System.out.println("3. Retirar dinero");
        System.out.println("4. Salir");
    }

    private static int leerEntero(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String linea = scanner.nextLine().trim();
            try {
                return Integer.parseInt(linea);
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un número válido.");
            }
        }
    }

    private static double leerDoublePositivo(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String linea = scanner.nextLine().trim();
            try {
                double valor = Double.parseDouble(linea);
                if (valor <= 0) {
                    System.out.println("Debe ingresar un monto mayor que cero.");
                } else {
                    return valor;
                }
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un número válido.");
            }
        }
    }

    private static String leerTextoNoVacio(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String texto = scanner.nextLine().trim();
            if (!texto.isEmpty()) {
                return texto;
            }
            System.out.println("El texto no puede estar vacío.");
        }
    }

    private static void realizarDeposito(Scanner scanner, CuentaUsuario cuenta) {
        double monto = leerDoublePositivo(scanner, "Monto a depositar: $");
        cuenta.depositar(monto);
        System.out.printf("Depósito exitoso. Nuevo saldo: $%.2f%n", cuenta.getSaldo());
    }

    private static void realizarRetiro(Scanner scanner, CuentaUsuario cuenta) {
        double monto = leerDoublePositivo(scanner, "Monto a retirar: $");
        if (cuenta.retirar(monto)) {
            System.out.printf("Retiro exitoso. Nuevo saldo: $%.2f%n", cuenta.getSaldo());
        } else {
            System.out.println("Fondos insuficientes. Operación cancelada.");
        }
    }

    private static Map<String, CuentaUsuario> cargarCuentas() {
        Map<String, CuentaUsuario> cuentas = new HashMap<>();
        File archivo = new File(ARCHIVO_CUENTAS);
        if (!archivo.exists()) {
            return cuentas;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(archivo))) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                String[] partes = linea.split("\\|");
                if (partes.length == 2) {
                    String usuario = partes[0];
                    double saldo = Double.parseDouble(partes[1]);
                    cuentas.put(usuario, new CuentaUsuario(usuario, saldo));
                }
            }
        } catch (IOException | NumberFormatException e) {
            System.out.println("Error al cargar cuentas: " + e.getMessage());
        }
        return cuentas;
    }

    private static void guardarCuentas(Map<String, CuentaUsuario> cuentas) {
        File archivo = new File(ARCHIVO_CUENTAS);
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(archivo))) {
            for (CuentaUsuario cuenta : cuentas.values()) {
                writer.write(cuenta.toRegistro());
                writer.newLine();
            }
        } catch (IOException e) {
            System.out.println("Error al guardar cuentas: " + e.getMessage());
        }
    }
}

class CuentaUsuario {

    private final String usuario;
    private double saldo;

    public CuentaUsuario(String usuario, double saldo) {
        this.usuario = usuario;
        this.saldo = saldo;
    }

    public String getUsuario() {
        return usuario;
    }

    public double getSaldo() {
        return saldo;
    }

    public void depositar(double monto) {
        if (monto > 0) {
            saldo += monto;
        }
    }

    public boolean retirar(double monto) {
        if (monto > 0 && monto <= saldo) {
            saldo -= monto;
            return true;
        }
        return false;
    }

    public String toRegistro() {
        return usuario + "|" + saldo;
    }
}

File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio 7.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <limits>

const std::string ARCHIVO_CUENTAS = "cuentas.txt";

class CuentaUsuario {
public:
    CuentaUsuario(const std::string& usuario, double saldo)
        : usuario(usuario), saldo(saldo) {}

    const std::string& getUsuario() const { return usuario; }
    double getSaldo() const { return saldo; }

    void depositar(double monto) {
        if (monto > 0) {
            saldo += monto;
        }
    }

    bool retirar(double monto) {
        if (monto > 0 && monto <= saldo) {
            saldo -= monto;
            return true;
        }
        return false;
    }

    std::string toRegistro() const {
        std::ostringstream out;
        out << usuario << "|" << saldo;
        return out.str();
    }

private:
    std::string usuario;
    double saldo;
};

int leerEntero(const std::string& mensaje) {
    while (true) {
        std::cout << mensaje;
        int valor;
        if (std::cin >> valor) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return valor;
        }
        std::cout << "Debe ingresar un número válido." << std::endl;
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }
}

double leerDoublePositivo(const std::string& mensaje) {
    while (true) {
        std::cout << mensaje;
        double valor;
        if (std::cin >> valor) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            if (valor > 0) {
                return valor;
            }
            std::cout << "Debe ingresar un monto mayor que cero." << std::endl;
        } else {
            std::cout << "Debe ingresar un número válido." << std::endl;
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        }
    }
}

std::string leerTextoNoVacio(const std::string& mensaje) {
    while (true) {
        std::cout << mensaje;
        std::string texto;
        std::getline(std::cin, texto);
        if (!texto.empty()) {
            return texto;
        }
        std::cout << "El texto no puede estar vacío." << std::endl;
    }
}

std::vector<CuentaUsuario> cargarCuentas() {
    std::vector<CuentaUsuario> cuentas;
    std::ifstream archivo(ARCHIVO_CUENTAS);
    if (!archivo.is_open()) {
        return cuentas;
    }

    std::string linea;
    while (std::getline(archivo, linea)) {
        std::istringstream ss(linea);
        std::string usuario;
        std::string saldoStr;
        if (std::getline(ss, usuario, '|') && std::getline(ss, saldoStr)) {
            try {
                double saldo = std::stod(saldoStr);
                cuentas.emplace_back(usuario, saldo);
            } catch (...) {
                // ignorar línea malformada
            }
        }
    }
    return cuentas;
}

void guardarCuentas(const std::vector<CuentaUsuario>& cuentas) {
    std::ofstream archivo(ARCHIVO_CUENTAS);
    if (!archivo.is_open()) {
        std::cout << "Error al guardar cuentas." << std::endl;
        return;
    }

    for (const auto& cuenta : cuentas) {
        archivo << cuenta.toRegistro() << '\n';
    }
}

CuentaUsuario* buscarCuenta(std::vector<CuentaUsuario>& cuentas, const std::string& usuario) {
    for (auto& cuenta : cuentas) {
        if (cuenta.getUsuario() == usuario) {
            return &cuenta;
        }
    }
    return nullptr;
}

void mostrarMenu() {
    std::cout << "=== Cajero Automático Básico ===" << std::endl;
    std::cout << "1. Consultar saldo" << std::endl;
    std::cout << "2. Depositar dinero" << std::endl;
    std::cout << "3. Retirar dinero" << std::endl;
    std::cout << "4. Salir" << std::endl;
}

int main() {
    std::vector<CuentaUsuario> cuentas = cargarCuentas();
    std::cout << "Bienvenido al Cajero Automático Básico." << std::endl;
    std::string usuario = leerTextoNoVacio("Ingrese su nombre de usuario: ");

    CuentaUsuario* cuenta = buscarCuenta(cuentas, usuario);
    if (cuenta == nullptr) {
        cuentas.emplace_back(usuario, 0.0);
        cuenta = &cuentas.back();
        std::cout << "Cuenta nueva creada con saldo inicial de $0.00." << std::endl;
    }

    int opcion;
    do {
        mostrarMenu();
        opcion = leerEntero("Seleccione una opción: ");
        switch (opcion) {
            case 1:
                std::cout << "Saldo actual de " << cuenta->getUsuario() << ": $" << cuenta->getSaldo() << std::endl;
                break;
            case 2: {
                double monto = leerDoublePositivo("Monto a depositar: $");
                cuenta->depositar(monto);
                guardarCuentas(cuentas);
                std::cout << "Depósito exitoso, nuevo saldo: $" << cuenta->getSaldo() << std::endl;
                break;
            }
            case 3: {
                double monto = leerDoublePositivo("Monto a retirar: $");
                if (cuenta->retirar(monto)) {
                    guardarCuentas(cuentas);
                    std::cout << "Retiro exitoso, nuevo saldo: $" << cuenta->getSaldo() << std::endl;
                } else {
                    std::cout << "Fondos insuficientes." << std::endl;
                }
                break;
            }
            case 4:
                std::cout << "Gracias por usar el cajero. Hasta luego." << std::endl;
                break;
            default:
                std::cout << "Opción inválida. Intente de nuevo." << std::endl;
        }
        std::cout << std::endl;
    } while (opcion != 4);

    return 0;
}
